File: iOSTutorial/iOSTutorial/SampleKit.swift
//
//  SampleNode.swift
//  SampleKit
//
//  Created by sangmin park on 1/31/24.
//

import Foundation
import iOSPubSub

class SampleKit{
    
    let messenger: Messenger
    
    init() {
        messenger = Messenger()
        messenger.setReceivingRule(all: Tag.native)
        messenger.setBasePublishingTag(Tag.game)
        messenger.subscribe(key: "test", handler: onTest)
        messenger.subscribe(key: "testRecall", handler: onTestRecall)
    }
    
    func onTest(message: Message){
        print("onTest : " + message.key);
        
        var container = Container()
        container.add(key: "data", value: "this is iOS message :D")
        let message = Message(key: "native", container: container)
        messenger.publish(message: message, tag: Tag.game)
    }
    func onTestRecall(message: Message){
        let data = message.container.getString(key: "data") ?? "no data....?"
        print("onTestRecall : " + data);
        
        var container = Container()
        container.add(key: "data", value: ("RECALL iOS => " + data))
        let reply = Message(key: "testReturn", container: container)
        messenger.publish(message: reply)
    }
    
    
}

File: AndroidTutorial/tutorial/src/main/java/com/pj/tutorial/SampleKit.kt
package com.pj.tutorial

import android.util.Log
import com.pj.pubsub.Messenger
import com.pj.pubsub.Tag
import com.pj.pubsub.extensions.ContainerBuilder
import com.pj.pubsub.extensions.Message
import com.pj.pubsub.extensions.add
import com.pj.pubsub.extensions.getString

class SampleKit {
    private val TAG = SampleKit::class.java.name

    private val messenger : Messenger = Messenger()

    init {
        messenger.apply {  }
        messenger.setReceivingRule(Tag.native)
        messenger.setBasePublishingTag(Tag.game)
        messenger.subscribe("test", this::onTest)
        messenger.subscribe("testRecall", this::onTestRecall)
    }

    private fun onTest(message: Message){
        Log.d(TAG, "onTest : $message" )

        val containerBuilder = ContainerBuilder()
        containerBuilder.add("data", "this is android message :D")
        val reply = Message("native", containerBuilder.build())
        messenger.publish(reply)
    }

    private fun onTestRecall(message: Message){
        val data = message.container.getString("data")
        Log.d(TAG, "onTestReCall : $data")

        val containerBuilder = ContainerBuilder()
        containerBuilder.add("data", "RECALL [$data]")
        val returned = Message("testReturn", containerBuilder.build())
        messenger.publish(returned)
    }
}